### PR TITLE
Bsc#1070961 corosync qdevice support

### PR DESCRIPTION
**Pull request by nick-wang**
Support corosync qdevice function, which will be a TP feature in SLE15.
(Please don't squash the commits, since possible to hide the UI info of corosync qdevice based on the feedback in future.)

### CODE REVIEW

File: package/cluster.fwd [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # 5560 for mgmtd
 # 7630 for hawk or hawk2
 # 21064 for dlm
+# 5403 for corosync qdevice(default)
```
Comment (jreidinger) [inline package/cluster.fwd]: it is nice that you document it, but I miss it in variable below.

File: src/include/cluster/dialogs.rb [hunk 2]
```diff
@@ -708,6 +708,205 @@ def SaveSecurity
...
+
+      if UI.QueryWidget(Id(:qdevice_port), :Value).to_i <= 0
+        Popup.Message(_("The corosync qdevice port must be a positive integer"))
+        UI.SetFocus(Id(:qdevice_port))
+        return false
+      end
+
+      if !["ffsplit", "lms", "test", "2nodelms"].include?(UI.QueryWidget(Id(:qdevice_algorithm), :Value))
+        Popup.Message(_("The algorithm only can be one of the ffsplit, lms, test or 2nodelms.
+          YaST will overwrite test and 2nodelms."))
```
Comment (jreidinger) [inline src/include/cluster/dialogs.rb]: for longer lines it is better to not have garbage whitespaces in string so you can write it this way:
```
 Popup.Message(_("The algorithm only can be one of the ffsplit, lms, test or 2nodelms." \
￼   " YaST will overwrite test and 2nodelms."))
```

File: src/include/cluster/dialogs.rb [hunk 3]
```diff
@@ -708,6 +708,205 @@ def SaveSecurity
...
+        UI.SetFocus(Id(:algorithm))
+        return false
+      end
+
+      if !["lowest", "highest"].include?(UI.QueryWidget(Id(:qdevice_tie_breaker), :Value)) &&
+        (UI.QueryWidget(Id(:tie_breaker), :Value).to_i <= 0)
+        Popup.Message(_("The tie breaker can be one of lowest, highest or a valid node id (number)"))
```
Comment (jreidinger) [inline src/include/cluster/dialogs.rb]: I am a bit confused here, widget is `qdevice_tie_breaker` or `tie_breaker`? you query both.

File: src/include/cluster/dialogs.rb [hunk 4]
```diff
@@ -708,6 +708,205 @@ def SaveSecurity
...
+      Cluster.corosync_qdevice = Convert.to_boolean(UI.QueryWidget(Id(:corosync_qdevice), :Value))
+
+      Cluster.qdevice_model = UI.QueryWidget(Id(:qdevice_model), :Value)
+      Cluster.qdevice_votes = UI.QueryWidget(Id(:qdevice_votes), :Value).to_s
+
+      Cluster.qdevice_host = UI.QueryWidget(Id(:qdevice_host), :Value)
+      Cluster.qdevice_port = UI.QueryWidget(Id(:qdevice_port), :Value).to_s
+      Cluster.qdevice_tls = UI.QueryWidget(Id(:qdevice_tls), :Value)
+      Cluster.qdevice_algorithm = UI.QueryWidget(Id(:qdevice_algorithm), :Value)
+      Cluster.qdevice_tie_breaker = UI.QueryWidget(Id(:qdevice_tie_breaker), :Value)
```
Comment (jreidinger) [inline src/include/cluster/dialogs.rb]: and here `qdevice_tie_breaker` is stored and not without prefix.

File: src/include/cluster/dialogs.rb [hunk 5]
```diff
@@ -708,6 +708,205 @@ def SaveSecurity
...
+        ),
+        HBox(
+          Left(ComboBox(
+            Id(:qdevice_tls), Opt(:hstretch), _("TLS:"),
+            ["off", "on", "required"]
+          )),
+          Left(ComboBox(
+            Id(:qdevice_algorithm), Opt(:hstretch, :notify), _("Algorithm:"),
+            [
+            Item(Id("ffsplit"),"ffsplit"),
```
Comment (jreidinger) [inline src/include/cluster/dialogs.rb]: indent here please

File: src/include/cluster/wizards.rb [hunk 6]
```diff
@@ -101,13 +102,18 @@ def FirstRunSequence
       sequence = {
         "ws_start"      => "communication",
         "communication" => {
+          :next  => "corosyncqdevice",
+          :back  => "communication",
```
Comment (jreidinger) [inline src/include/cluster/wizards.rb]: this looks strange, having back to itself. It back needed here at all? maybe it should be just disabled.

File: src/include/cluster/dialogs.rb [hunk 7]
```diff
@@ -708,6 +708,205 @@ def SaveSecurity
...
+
+      UI.ChangeWidget(Id(:qdevice_model), :Value, Cluster.qdevice_model)
+      UI.ChangeWidget(Id(:qdevice_votes), :Value, Cluster.qdevice_votes)
+      UI.ChangeWidget(Id(:qdevice_votes), :ValidChars, "0123456789");
+
+      UI.ChangeWidget(Id(:qdevice_host), :Value, Cluster.qdevice_host)
+      UI.ChangeWidget(Id(:qdevice_port), :Value, Cluster.qdevice_port)
+      UI.ChangeWidget(Id(:qdevice_tls), :Value, Cluster.qdevice_tls)
+      UI.ChangeWidget(Id(:qdevice_algorithm), :Value, Cluster.qdevice_algorithm)
+      UI.ChangeWidget(Id(:qdevice_tie_breaker), :Value, Cluster.qdevice_tie_breaker)
```
Comment (jreidinger) [inline src/include/cluster/dialogs.rb]: what about having init method for whole this initialization?

File: package/cluster.fwd (hunk 1 above)
Comment (nick-wang) [reply package/cluster.fwd]: This port is configurable in UI and default value is 5403. Will set the actual port via configuration.
"5403" Refer to "src/include/cluster/dialogs.rb" for drawing UI and "src/modules/ag_openais" for read from file.

File: src/include/cluster/dialogs.rb (hunk 2 above)
Comment (nick-wang) [reply src/include/cluster/dialogs.rb]: Thanks. Remember it.

File: src/include/cluster/dialogs.rb (hunk 3 above)
Comment (nick-wang) [reply src/include/cluster/dialogs.rb]: oops, it is a bug. I changed the label to :qdevice_tie_breaker but missing two of them.
Thanks for the catch.

File: src/include/cluster/dialogs.rb (hunk 4 above)
Comment (nick-wang) [reply src/include/cluster/dialogs.rb]: Yes, should change all :tie_breaker to :qdevice_tie_breaker.

File: src/include/cluster/dialogs.rb (hunk 5 above)
Comment (nick-wang) [reply src/include/cluster/dialogs.rb]: Thanks.

File: src/include/cluster/dialogs.rb (hunk 7 above)
Comment (nick-wang) [reply src/include/cluster/dialogs.rb]: Move the code to function CorosyncQdeviceLayout(). It is mainly about initialization, including draw the UI and initialize the parameters.

File: src/include/cluster/wizards.rb (hunk 6 above)
Comment (nick-wang) [reply src/include/cluster/wizards.rb]: “Back” button still appear when removing line 106. So leave the code without change.

**Comment by nick-wang** [comment]
Squash and rebase the fix commit into the last commit.

**Comment by nick-wang** [comment]
@jreidinger Could you help to review the changes? thanks.

**Pull request merged**